File: hls_proxy.py
#!/usr/bin/env python3
"""HLS 代理服务器（多线程）：4000 端口 → /hls/* → MediaMTX :8888，/* → 静态文件"""
import http.server
import socketserver
import urllib.request
import os
import sys

PORT = 4000
TARGET_HOST = 'http://localhost:8888'

class ThreadingHTTPServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
    """多线程 HTTP 服务器，支持并发请求"""
    daemon_threads = True

class HLSProxy(http.server.SimpleHTTPRequestHandler):
    def do_GET(self):
        if self.path.startswith('/hls/'):
            hls_path = self.path[4:]  # 去掉 /hls 前缀
            url = f'{TARGET_HOST}{hls_path}'
            try:
                req = urllib.request.Request(url, headers={
                    'User-Agent': 'HLS-Proxy/1.0'
                })
                with urllib.request.urlopen(req, timeout=5) as resp:
                    body = resp.read()
                    self.send_response(200)
                    ct = resp.headers.get('Content-Type', 'application/octet-stream')
                    self.send_header('Content-Type', ct)
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.send_header('Cache-Control', 'no-store')
                    self.send_header('Content-Length', str(len(body)))
                    self.end_headers()
                    self.wfile.write(body)
            except Exception as e:
                self.send_response(502)
                self.send_header('Content-Type', 'text/plain')
                self.end_headers()
                self.wfile.write(f'HLS proxy error: {e}'.encode())
        elif self.path == '/':
            self.send_response(302)
            self.send_header('Location', '/yolov26-rtsp-detection.html')
            self.end_headers()
        else:
            super().do_GET()
    
    def log_message(self, format, *args):
        pass  # 静默日志，减少输出

if __name__ == '__main__':
    os.chdir('/Users/paocai/WorkBuddy/2026-05-19-task-13')
    httpd = ThreadingHTTPServer(('', PORT), HLSProxy)
    print(f'HLS Proxy (multi-threaded) on http://localhost:{PORT}')
    sys.stdout.flush()
    httpd.serve_forever()

File: hls_proxy.js
#!/usr/bin/env node
/**
 * HLS 代理服务器 (Node.js)
 * 端口 4000: 静态文件 + HLS 流代理到 MediaMTX :8888
 */
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 4000;
const ROOT = '/Users/paocai/WorkBuddy/2026-05-19-task-13';
const HLS_TARGET = 'http://127.0.0.1:8888';

function contentType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const types = {
    '.html': 'text/html; charset=utf-8',
    '.css': 'text/css',
    '.js': 'application/javascript',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.mp4': 'video/mp4',
    '.m3u8': 'application/vnd.apple.mpegurl',
    '.ts': 'video/mp2t',
    '.mp3': 'audio/mpeg',
  };
  return types[ext] || 'application/octet-stream';
}

const server = http.createServer((req, res) => {
  const url = req.url.split('?')[0];

  // HLS 代理：转发到 MediaMTX
  if (url.startsWith('/hls/')) {
    const targetPath = url.slice(4); // 去掉 /hls
    const targetUrl = `${HLS_TARGET}${targetPath}`;

    const proxy = http.get(targetUrl, (proxyRes) => {
      res.writeHead(proxyRes.statusCode, {
        'Content-Type': proxyRes.headers['content-type'] || 'application/octet-stream',
        'Access-Control-Allow-Origin': '*',
        'Cache-Control': 'no-store',
      });
      proxyRes.pipe(res);
    });

    proxy.on('error', (err) => {
      res.writeHead(502);
      res.end('HLS proxy error: ' + err.message);
    });

    proxy.setTimeout(30000, () => {
      proxy.destroy();
      res.writeHead(504);
      res.end('HLS proxy timeout');
    });

    return;
  }

  // 首页重定向
  if (url === '/') {
    res.writeHead(302, { 'Location': '/yolov26-rtsp-detection.html' });
    res.end();
    return;
  }

  // 静态文件
  const filePath = path.join(ROOT, url);
  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404);
      res.end('Not Found');
      return;
    }
    res.writeHead(200, { 'Content-Type': contentType(filePath), 'Cache-Control': 'no-cache' });
    res.end(data);
  });
});

server.listen(PORT, () => {
  console.log(`HLS Proxy (Node.js) on http://localhost:${PORT}`);
  console.log(`  /hls/*  →  ${HLS_TARGET}/*`);
  console.log(`  /*      →  static files from ${ROOT}`);
});
